BJS2-84316-recommend_analytics

Linempy opened the pull request:
Задача "Аналитика рекомендаций"

Linempy committed:
BJS2-84316-recommend_analytics
- написал аналитику рекомендаций
- написал тесты

Linempy committed:
BJS2-84316-recommend_analytics
- немного подкорректировал код

RinatBeybutov commented on src/main/java/faang/school/analytics/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,85 @@
+                = new RedisMessageListenerContainer();
+        container.setConnectionFactory(jedisConnectionFactory());
+        container.addMessageListener(listener, topic());
+        log.info("ДОЛЖЕН ВЫЗЫВАТЬСЯ!");
```
что за странный лог

RinatBeybutov commented on src/main/java/faang/school/analytics/dto/RecommendationFilterDto.java:
```diff
@@ -0,0 +1,38 @@
+        LocalDateTime startTime,
+        LocalDateTime endTime
+) {
+    public void validate() {
```
посмотри вариант с @AssertTrue, может так лучше будет

RinatBeybutov commented on src/main/java/faang/school/analytics/model/EventTypeRegistry.java:
```diff
@@ -0,0 +1,31 @@
+        EVENT_TYPE_MAP.put(RecommendationEvent.class , EventType.RECOMMENDATION_RECEIVED);
+    }
+
+    public static EventType getEventTypeForClass(Class<?> eventClass) {
```
а что мешает эту функциональность в enum запихнуть? а так у тебя лишний класс на ровном месте

RinatBeybutov commented on src/main/java/faang/school/analytics/model/TimeIntervalType.java:
```diff
@@ -0,0 +1,55 @@
+ * @since 20.08.2025
+ */
+public enum TimeIntervalType {
+    DAY(d -> d.toLocalDate().atStartOfDay(),
```
что-то очень громоздко выглядит, подумай, как это сократить можно

RinatBeybutov commented on src/main/java/faang/school/analytics/repository/AnalyticsEventRepository.java:
```diff
@@ -1,14 +1,21 @@
 
-    Stream<AnalyticsEvent> findByReceiverIdAndEventType(long receiverId, EventType eventType);
+    default List<AnalyticsEvent> findByFilter(RecommendationFilterDto filterDto) {
+        Specification<AnalyticsEvent> filter = SpecificationBuilder.buildSpecification(filterDto);
```
давай не будем создавать дефолтный метод, а спеку и поиск будем писать в сервисе

RinatBeybutov commented on src/main/java/faang/school/analytics/repository/specification/SpecificationBuilder.java:
```diff
@@ -0,0 +1,73 @@
+public class SpecificationBuilder {
+
+    public static Specification<AnalyticsEvent> buildSpecification(RecommendationFilterDto filterDto) {
+        return Specification.where(byReceiverId(filterDto.id()))
```
можно заменить на Specification.allOf

RinatBeybutov commented on src/main/java/faang/school/analytics/service/AnalyticsEventServiceImpl.java:
```diff
@@ -0,0 +1,35 @@
+    private final AnalyticsEventRepository repository;
+
+    @Override
+    public void saveEvent(AnalyticsEvent event) {
```
транзакцию забыл

RinatBeybutov commented on src/test/java/faang/school/analytics/controller/AnalyticsEventControllerTest.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @ParameterizedTest
+    @MethodSource("faang.school.analytics.controller.AnalyticsEventControllerData#validFilterDto")
```
поправь форматирование стиля

RinatBeybutov commented on src/test/java/faang/school/analytics/service/AnalyticsEventServiceImplTest.java:
```diff
@@ -0,0 +1,75 @@
+    private AnalyticsEventRepository repository;
+
+    @InjectMocks
+    AnalyticsEventServiceImpl service;
```
private

RinatBeybutov commented:
Напиши хотя бы один тест с использованием testcontainers интеграционный

Linempy committed:
BJS2-84316-recommend_analytics
- написал 2 интеграционных теста

Linempy committed:
BJS2-84316-recommend_analytics
- исправил замечания

RinatBeybutov commented on src/main/java/faang/school/analytics/model/EventType.java:
```diff
@@ -18,6 +25,12 @@ public enum EventType {
+    private static final Map<Class<?>, EventType> EVENT_TYPE_MAP = new HashMap<>();
+
+    static {
+        EVENT_TYPE_MAP.put(RecommendationEvent.class , EventType.RECOMMENDATION_RECEIVED);
```
костыльно выглядит, давай в этом енаме создадим поле
private Class<?> linkedClass;

Добавим конструктор и тогда объявление перечислений будет выглядеть так:
..
RECOMMENDATION_RECEIVED(RecommendationEvent.class),
ADDED_TO_FAVOURITES(null),
...

А дальше получение по классу нужно поправить:
проходишься по всем значениям, если есть совпадение по классу - возвращаешь, а если за весь цикл не нашел, то кидай ошибку, что такой не поддерживается

RinatBeybutov commented on src/main/java/faang/school/analytics/repository/AnalyticsEventRepository.java:
```diff
@@ -1,25 +1,15 @@
-
 @Repository
-public interface AnalyticsEventRepository extends CrudRepository<AnalyticsEvent, Long>,
+public interface AnalyticsEventRepository extends JpaRepository<AnalyticsEvent, Long>,
```
аннотацию @Repository убери

RinatBeybutov commented on src/test/java/faang/school/analytics/controller/AnalyticsEventControllerTestIT.java:
```diff
@@ -52,9 +52,6 @@ public class AnalyticsEventControllerTestIT {
 
     @DynamicPropertySource
     static void postgresqlProperties(DynamicPropertyRegistry registry) {
-        POSTGRESQL_CONTAINER.start();
```
почему убрал?

RinatBeybutov commented on src/main/java/faang/school/analytics/controller/AnalyticsEventController.java:
```diff
@@ -0,0 +1,41 @@
+     * @return отфильтрованный список данных
+     */
+    @GetMapping
+    public ResponseEntity<List<AnalyticsEvent>> getByFilter(@Valid @ModelAttribute RecommendationFilterDto filterDto) {
```
entity на фронт отдаешь?

Linempy committed:
BJS2-84316-recommend_analytics
- поправил тесты
- убрал костыли в EventType
- создал AnalyticsViewDto
- поправил замечания

Linempy committed:
BJS2-84316-recommend_analytics
- изменил тип с LocalDateTime на LocalDate в TimeIntervalType

RinatBeybutov commented on src/main/java/faang/school/analytics/dto/RecommendationFilterDto.java:
```diff
@@ -0,0 +1,45 @@
+public class RecommendationFilterDto {
+
+    @NonNull
+    private Long id;
```
это не entity, поэтому лучше указать recommendationId

RinatBeybutov commented on src/main/java/faang/school/analytics/model/EventType.java:
```diff
@@ -26,4 +36,14 @@ public static EventType of(int type) {
+
+    public static EventType getEventTypeForClass(Class<?> eventClass) {
+        for (EventType eventType : EventType.values()) {
+            if (eventType.linkedClass == eventClass) {
```
пускай и работает, но лучше все-таки использовать equals, на будущее
